colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,445 @@
+        return update
+
+    def _base_factor(self, update, grad, state, grad_shape, beta2t):
+        if self.use_zero:
+            # only zero
+            if grad_shape[0] % self.data_parallel_size != 0:
+                # view update to origin shape update.view(grad_shape[0]//self.data_parallel_size , grad_shape[1])
+                # row mean no change
+                # col mean need reduce and div
+                # gather update[flatten] along dp group then reshape to [H, W]
+                update = _gather(input_=update, dim=-1, process_group=self.data_parallel_group)
+                # view update to origin[tp] shape
+                update_reshape = update.view(-1, grad_shape[1])
+                # gather grad[flatten] along dp group then reshape to [H, W]
+                grad = _gather(input_=grad, dim=-1, process_group=self.data_parallel_group)
+                grad_reshape = grad.view(-1, grad_shape[1])
+                exp_avg_sq_row = state["exp_avg_sq_row"]  # [H/dp]
+                exp_avg_sq_col = state["exp_avg_sq_col"]  # [W]
+                exp_avg_sq_row.mul_(beta2t).add_(update_reshape.mean(dim=-1), alpha=(1.0 - beta2t))
+                exp_avg_sq_col.mul_(beta2t).add_(update_reshape.mean(dim=-2), alpha=(1.0 - beta2t))
+                # reduce col
+                dist.all_reduce(exp_avg_sq_col, group=self.tensor_parallel_group)
+                exp_avg_sq_col.div_(self.tensor_parallel_size)
+                update_reshape = self._approx_sq_grad(exp_avg_sq_row, exp_avg_sq_col)
+                update_reshape.mul_(grad_reshape)
+                update = _split(input_=update_reshape.view(-1), dim=-1, process_group=self.data_parallel_group)
+            else:
+                # no residual row
+                # view update to origin[tp] shape
+                update_reshape = update.view(-1, grad_shape[1])  # [H/dp, W]
+                grad_reshape = grad.view(-1, grad_shape[1])  # [H/dp, W]
+                # print(f"grad_shape {grad_shape} update shape {update.shape} grad shape {grad.shape}\n update {update}\n")
+                exp_avg_sq_row = state["exp_avg_sq_row"]  # [H/tp]
+                exp_avg_sq_col = state["exp_avg_sq_col"]  # [W]
+                exp_avg_sq_row.mul_(beta2t).add_(update_reshape.mean(dim=-1), alpha=(1.0 - beta2t))
+                exp_avg_sq_col.mul_(beta2t).add_(update_reshape.mean(dim=-2), alpha=(1.0 - beta2t))
+                # reduce col
+                dist.all_reduce(exp_avg_sq_col, group=self.tensor_parallel_group)
+                exp_avg_sq_col.div_(self.tensor_parallel_size)
+                update_reshape = self._approx_sq_grad(exp_avg_sq_row, exp_avg_sq_col)
+                update_reshape.mul_(grad_reshape)
+                update = update_reshape.view(-1)
+                # print(f"No res factor exp_avg_sq_col is_dtensor {False} shard_spec {None} use_zero {self.use_zero} dp_size {self.data_parallel_size} tp_size {self.tensor_parallel_size}\n {state['exp_avg_sq_col']}\n")
```

Comment:
Remove such meaningless comments

Reply:
Removed in https://github.com/hpcaitech/ColossalAI/pull/5484/commits/1038b233632c5ad04c40262a7027ba7205a14e65;